👤 feat: User Placeholder Variables for Custom Endpoint Headers

danny-avila opened the pull request.

danny-avila committed:
🔧 refactor: move `processMCPEnv` from `librechat-data-provider` and move to `@librechat/api`

owengo commented:
Thank you for this, this will be very useful for mcp integration with https://github.com/taylorwilsdon/google_workspace_mcp for example, we won't rely on the llm to identify users and prevent users spoofing other accounts

danny-avila commented:
@owengo this is already available for MCP servers:
https://www.librechat.ai/docs/configuration/librechat_yaml/object_structure/mcp_servers#mcp-server-with-dynamic-user-fields

owengo commented:
It goes so fast! It was still not merged on my branch.
We are still missing the Authorization header, it would allow google credential injection on remote MCPs: for example:   https://github.com/taylorwilsdon/google_workspace_mcp/pull/72
This way the oauth process, user-wise, would happen 100% on librechat auth and the mcps giving access to google drive etc would just use the tokens passed by librechat ( currently it's quite clumsy becaus the mcp needs a tool to generate an oauth link and convince the user to click it .. )

dustinhealy committed:
🔧 refactor: Update resolveHeaders import paths

dustinhealy committed:
🔧 refactor: Enhance resolveHeaders to support user and custom variables

- Updated resolveHeaders function to accept user and custom user variables for placeholder replacement.
- Modified header resolution in multiple client and controller files to utilize the enhanced resolveHeaders functionality.
- Added comprehensive tests for resolveHeaders to ensure correct processing of user and custom variables.

dustinhealy committed:
🔧 fix: Update user ID placeholder processing in env.ts

dustinhealy committed:
🔧 fix: Remove arguments passing this.user rather than req.user

- Updated multiple client and controller files to call resolveHeaders without the user parameter

danny-avila commented on packages/api/src/utils/env.ts:
```diff
@@ -39,8 +40,21 @@ function processUserPlaceholders(value: string, user?: TUser): string {
+      // May be unnecessary, currently here just to pass mcp.spec.ts test: should support both id and _id properties for LIBRECHAT_USER_ID
+      if (field === 'id') {
+        if (fieldValue !== undefined && fieldValue !== '') {
+          const replacementValue = fieldValue != null ? String(fieldValue) : '';
```
code is repeated here, please apply DRY

dustinhealy committed:
🔧 refactor: Enhance processUserPlaceholders to be more readable / less nested

dustinhealy committed:
🔧 refactor: Update processUserPlaceholders to pass all tests in mpc.spec.ts and env.spec.ts

danny-avila committed:
chore: remove legacy ChatGPTClient

danny-avila committed:
chore: remove LLM initialization code

danny-avila committed:
chore: initial deprecation removal of `gptPlugins`

danny-avila committed:
chore: remove cohere-ai dependency from package.json and package-lock.json

danny-avila committed:
chore: update brace-expansion to version 2.0.2 and add license information

danny-avila committed:
chore: remove PluginsClient test file

danny-avila committed:
chore: remove legacy

danny-avila committed:
ci: remove deprecated sendMessage/getCompletion/chatCompletion tests
